### PR TITLE
Introduce torch_xla.launch()

### PR DESCRIPTION
**Pull request by zpcore**
Draft of using `torch_xla.launch()` for `torchrun` and `xmp.spawn()`

Unify `_mp_fn(xu.getenv_as(xenv.LOCAL_RANK, int), FLAGS)` (call with torchrun) or   
`xmp.spawn(_mp_fn, args=(FLAGS,), nprocs=FLAGS.num_cores)`  
with 
```
torch_xla.launch(_mp_fn, args=(FLAGS,) , nprocs=FLAGS.num_cores)
```

### CODE REVIEW

File: torch_xla/launch.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import torch.distributed as dist
+import torch_xla
+import torch_xla.core.xla_env_vars as xenv
+import torch_xla.distributed.xla_multiprocessing as xmp
+import torch_xla.utils.utils as xu
+import torch_xla.runtime as xr
+
+
+def spawn_wrapper(index, fn, *args, **kwargs):
+  dist.init_process_group('xla', init_method='xla://')
```
Comment (will-cromar) [inline torch_xla/launch.py:12]: PyTorch's documentation shows `init_process_group` inside of the user's main function, e.g. https://pytorch.org/tutorials/intermediate/dist_tuto.html

We should stick to that pattern and not `init_process_group` ourselves to be safe.

File: torch_xla/launch.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  fn(index, *args, **kwargs)
+
+
+def launch(
+    fn,
+    *args,
+    **kwargs,
```
Comment (will-cromar) [inline torch_xla/launch.py:19]: `spawn` does not actually take `kwargs`, so for consistency, you should not take `kwargs` for `fn` in this wrapper. https://pytorch.org/docs/stable/multiprocessing.html#spawning-subprocesses

Also, I don't recommend mixing variable-length arguments for both `launch` and `fn`. You can get around that by making `args` a singular tuple, similar to `spawn`.

File: torch_xla/launch.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+  Raises:
+    NotImplementedError: SPMD is not supported yet.
+  """
+  if xr.is_spmd():
+    # TODO(piz): SPMD is specified differently from mp. Skip for now.
+    raise NotImplementedError(
+        'launch function does not support SPMD at this time')
+
+  nprocs = kwargs.pop('nprocs', None)
```
Comment (will-cromar) [inline torch_xla/launch.py:31]: Put only the options that are implemented here directly as keyword arguments. Namely, `spawn_method` and `nprocs`

The only valid states of `nprocs` are 1 and `num_devices`. You can make the illegal states impossible by replacing it with a bool (e.g. `debug_single_process`) that determines whether we pass `1` to `xmp.spawn` or not.

`join` and `daemon` are not implemented in `xmp.spawn` and were just there to mirror torch's `spawn`.

File: examples/data_parallel/train_resnet_ddp.py [hunk 4]
```diff
@@ -26,5 +27,7 @@ def _mp_fn(index):
 
 
 if __name__ == '__main__':
-  print('consider using train_resnet_spmd_data_parallel.py instead to get better performance')
-  xmp.spawn(_mp_fn, args=())
+  print(
+      'consider using train_resnet_spmd_data_parallel.py instead to get better performance'
+  )
+  torch_xla.launch(_mp_fn, nprocs=1)
```
Comment (will-cromar) [inline examples/data_parallel/train_resnet_ddp.py:33]: Why `nprocs=1`?

File: torch_xla/launch.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import torch_xla.utils.utils as xu
+import torch_xla.runtime as xr
+
+
+def spawn_wrapper(index, fn, *args, **kwargs):
+  dist.init_process_group('xla', init_method='xla://')
+  fn(index, *args, **kwargs)
+
+
+def launch(
```
Comment (will-cromar) [inline torch_xla/launch.py:16]: nit: put this in `torch_xla.py` with the other top-level functions if you can

File: examples/data_parallel/train_resnet_ddp.py (hunk 4 above)
Comment (zpcore) [reply examples/data_parallel/train_resnet_ddp.py:33]: I can leave it to nprocs=None as it is before.

File: torch_xla/launch.py (hunk 3 above)
Comment (zpcore) [reply torch_xla/launch.py:31]: Regarding `The only valid states of nprocs are 1 and num_devices` , is this also true for GPU?

File: torch_xla/launch.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  """
+  if xr.is_spmd():
+    # TODO(piz): SPMD is specified differently from mp. Skip for now.
+    raise NotImplementedError(
+        'launch function does not support SPMD at this time')
+
+  if debug_single_process:
+    nprocs = 1
+  else:
+    nprocs = xr.global_runtime_device_count()
```
Comment (will-cromar) [inline torch_xla/launch.py:36]: Set this to `None` and `spawn` will do the right thing without initializing the runtime

File: torch_xla/launch.py (hunk 2 above)
Comment (zpcore) [reply torch_xla/launch.py]: Cleaned up, it's now much simpler.

File: examples/data_parallel/train_resnet_ddp.py [hunk 7]
```diff
@@ -8,13 +8,15 @@
...
+import torch_xla
+import torch_xla.core.xla_model as xm
 
 
 class TrainResNetDDP(TrainResNetBase):
 
   def __init__(self):
-    super().__init__()
     dist.init_process_group('xla', init_method='xla://')
+    super().__init__()
```
Comment (will-cromar) [inline examples/data_parallel/train_resnet_ddp.py:19]: nit: should the super init stay at the top?

File: torch_xla/torch_xla.py [hunk 8]
```diff
@@ -80,3 +85,27 @@ def manual_seed(seed, device=None):
       If missing the default device seed will be set.
   """
   xm.set_rng_state(seed, device)
+
+
+def launch(
+    fn: Callable,
+    args: Tuple = (),
```
Comment (will-cromar) [inline torch_xla/torch_xla.py:92]: Can you add a TODO to make `args` a ParamSpec? I want to take a pass to improve our type annotations, but it'll involve adding `typing` backports as a dependency. `Tuple` is correct and works for this PR.

https://docs.python.org/3/library/typing.html#typing.ParamSpec

**Review by will-cromar — APPROVED** [review]
Nice! Thank you for implementing this :smile:

File: torch_xla/torch_xla.py (hunk 8 above)
Comment (zpcore) [reply torch_xla/torch_xla.py:92]: Oh, I just pressed the merge without noticing the comment. Will update in another PR.

File: examples/data_parallel/train_resnet_ddp.py (hunk 7 above)
Comment (zpcore) [reply examples/data_parallel/train_resnet_ddp.py:19]: No, super().__init__() will call some functions from runtime, which needs init the TPU backend first.

**Pull request merged**